crates/runtime/src/instance/allocator/pooling.rs

```diff
@@ -736,161 +310,168 @@ impl PoolingInstanceAllocator {
-            }
+        if num_tables > usize::try_from(self.limits.max_tables_per_component).unwrap() {
+            bail!(
+                "The component transitively contains {num_tables} tables, which exceeds the \
+                 configured maximum of {}",
+                self.limits.max_tables_per_component
+            );
         }
-    }
 
-    fn allocate_tables(
-        &self,
-        index: usize,
-        req: &mut InstanceAllocationRequest,
-        tables: &mut PrimaryMap<DefinedTableIndex, Table>,
-    ) -> Result<()> {
-        let module = req.runtime_info.module();
+        Ok(())
+    }
 
+    fn validate_module_impl(&self, module: &Module, offsets: &VMOffsets<HostPtr>) -> Result<()> {
+        self.validate_memory_plans(module)?;
         self.validate_table_plans(module)?;
+        self.validate_core_instance_size(offsets)?;
+        Ok(())
+    }
 
-        let mut bases = self.tables.get(index);
-        for (_, plan) in module.table_plans.iter().skip(module.num_imported_tables) {
-            let base = bases.next().unwrap() as _;
+    fn increment_component_instance_count(&self) -> Result<()> {
+        let old_count = self.live_component_instances.fetch_add(1, Ordering::AcqRel);
+        if old_count >= u64::from(self.limits.total_component_instances) {
+            self.decrement_component_instance_count();
+            bail!(
+                "maximum concurrent component instance limit of {} reached",
+                self.limits.total_component_instances
+            );
+        }
+        Ok(())
+    }
 
-            commit_table_pages(
-                base as *mut u8,
-                self.tables.max_elements as usize * mem::size_of::<*mut u8>(),
-            )?;
+    fn decrement_component_instance_count(&self) {
+        self.live_component_instances.fetch_sub(1, Ordering::AcqRel);
+    }
 
-            tables.push(Table::new_static(
-                plan,
-                unsafe { std::slice::from_raw_parts_mut(base, self.tables.max_elements as usize) },
-                unsafe { &mut *req.store.get().unwrap() },
-            )?);
+    fn increment_core_instance_count(&self) -> Result<()> {
+        let old_count = self.live_core_instances.fetch_add(1, Ordering::AcqRel);
+        if old_count >= u64::from(self.limits.total_core_instances) {
+            self.decrement_core_instance_count();
+            bail!(
+                "maximum concurrent core instance limit of {} reached",
+                self.limits.total_core_instances
+            );
         }
-
         Ok(())
     }
 
-    fn deallocate_tables(&self, index: usize, tables: &mut PrimaryMap<DefinedTableIndex, Table>) {
-        // Decommit any tables that were used
-        for (table, base) in tables.values_mut().zip(self.tables.get(index)) {
-            let table = mem::take(table);
-            assert!(table.is_static());
+    fn decrement_core_instance_count(&self) {
+        self.live_core_instances.fetch_sub(1, Ordering::AcqRel);
```

Comment:
Mind throwing in a debug assert here that the return value is not 0? (e.g. this never goes negative)

Reply:
(and the other decrement methods too)